### ISSUE TITLE
Indic decomposing / recomposing

### ISSUE THREAD

**Issue by tiroj** [harfbuzz/harfbuzz#3356]
I am trying to understand what is happening in HB handling of some Devanagari nukta letters, and whether any of the behaviour constitutes a bug. In a Twitter discussion on this question, @behdad wrote

> The HB Indic shaper first fully decomposes (as long as the font has glyphs for the characters), then tries to compose (again, as long as the font has glyphs for the characters), without shortcircuiting. The latter part is where it different from eg. the default shaper.

The behaviour I am seeing with some nukta letters is that decomposition is happening as described, but the second, (re)composition is not happening, even though the font definitely has glyphs for the composed characters. For input ऩ ऱ ऴ I was getting:

![image](https://user-images.githubusercontent.com/5614500/148425605-2b6976ae-6ca3-422a-832b-24a05699993c.png)

The dots on the ऩ ऴ are incorrectly positoned because the precomposed characters have been decomposed, not recomposed, and do not contain blwm anchor positioning for nukta (as is fairly common in Devanagari fonts targeting Hindi, in which the number of nukta letters is a small subset and all included in Unicode as precomposed characters).

[@behdad `[explained](https://twitter.com/behdadesfahbod/status/1479116066756313093)` why the Ra is not decomposed, which makes perfect sense.]

A workaround to this is, of course, to ensure that all the precomposed nukta letter glyphs are also mapped from decomposed sequences in the font’s nukt feature. Since discovering the HB behaviour yesterday, I have already updated the current font in this way, and will go back to older fonts and do the same.

However, I think there may still be a bug in HB’s handling, since I believe recomposition at the character level should apply here as @behdad said: I don’t think anything in recomposing these sequences would constitute a ‘shortcircuit’.

Confusing the issue a bit further is that HB _does_ recompose these nukta letters if a combining mark character is applied to any of them in a line, i.e. it recomposes all the instances of the nukta letters in that line, not just the one that takes the combining mark:
![image](https://user-images.githubusercontent.com/5614500/148426721-51f088f6-3844-48c9-9fb4-4758cc868811.png)

@dscorbett [suggested](https://twitter.com/d_s_corbett/status/1478880389129289729
) an optimisation was involved in this, but it doesn’t make a lot of sense to me, and I am thinking that probably HB should be trying to recompose the Unicode nukta letters whenever they occur.

**Comment by behdad** [comment]
A test font is needed. We can't help otherwise.

**Comment by tiroj** [comment]
Can test with current release version of Microsoft’s Sanskrit Text font: it maps most of the precomposed nukta letters in the nukt feature, but not ऩ and ऴ. I am working on updates to this font at the moment, and that’s how I stumbled across this shaping issue.

**Comment by behdad** [comment]
> Can test with current release version of Microsoft’s Sanskrit Text font

What's the file name?

**Comment by behdad** [comment]
> @dscorbett [suggested](https://twitter.com/d_s_corbett/status/1478880389129289729) an optimisation was involved in this, but it doesn’t make a lot of sense to me, and I am thinking that probably HB should be trying to recompose the Unicode nukta letters whenever they occur.

No that's not relevant here.

**Comment by behdad** [comment]
> Can test with current release version of Microsoft’s Sanskrit Text font

What's the file name?

**Comment by tiroj** [comment]
I think it is sanskr_.ttf

Later today, I can send you a test font derived from Tiro Devanagari Hindi with the nukt feature disabled. That will be useful for testing behaviour for all the precomposed Unicode nukta characters (and means I will also be able to confirm how all of them are handled in other shaping engines with both C and D normalisations).

**Comment by behdad** [comment]
> I think it is sanskr_.ttf

I can't find it in the Windows fonts I have access to. Perhaps it's in the Office bundle.

> Later today, I can send you a test font derived from Tiro Devanagari Hindi with the nukt feature disabled

Please do. I cannot help until someone sends me a reproduction font.

**Comment by tiroj** [comment]
Curiouser and curiouser:

I made a couple of test fonts, but am unable to reproduce the same bug as yesterday when inputting text directly in Chrome browser (testing via dropping fonts in axis-praxis.org) rather than copying and pasting text from my InDesign file as yesterday. Will explore further when I get back from walking the dog, but now wondering if the issue is related to clipboard contents.

**Comment by tiroj** [comment]
Well...

Today I am unable to reproduce the behaviour that I reported and illustrated on Thursday with the test fonts made for the purpose. I made two test fonts:

The TEST 1 font has the nukta OTL feature completely removed, so I could see the handling of these precomposed Unicode nukta characters क़ ख़ ग़ ज़ ड़ ढ़ ऩ फ़ य़ ऱ ऴ in different environments without glyph level processing, i.e. relying on text layout or shaping engines to perform or not perform normalisation. My expectation of correct behaviour is that these characters should display using the precomposed glyphs in the font, either mapped directly from undecomposed Unicode characters (or NFC) or from decomposed Unicode characters (NFD: क़ ख़ ग़ ज़ ड़ ढ़ ऩ फ़ य़ ऱ ऴ) recomposed by the shaping engine prior to glyph display.

The TEST 2 font retains the nukta OTL feature, but has its lookup edited to remove the composition mappings for ढ़ ऩ फ़ य़ ऱ, i.e. providing mapping only for क़ ख़ ग़ ज़ ड़. This enables me to see whether any environments treat these subsets differently, e.g. decomposing the nukta characters and always relying on the presence of the nukt feature to compose the nukta glyphs. Again, my expectation of correct display is that the precomposed glyphs should be used for all these nukta letters, on the basis of their Unicode canonical compositions, and the presence or absence of nukt feature lookups should make no difference.

I tested in current official release versions of Pages (Apple CoreText engine), Chrome browser (HarfBuzz), and InDesign 2022 (Adobe WRC _I think_), and both precomposed and decomposed character sequences are rendering correctly in all environments.

I'll go back to the earlier version of the Sanskrit Text font that exhibited the problem on Friday, and see if I can reproduce it with that. If am able to reproduce the issue, I will send that test font to @behdad
